http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -184,10 +187,58 @@ openSocket tweakSocket addr =
             NS.connect sock (NS.addrAddress addr)
             return sock)
 
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
```

Comment:
I'm overall having trouble following the intended logic of this code. If I understand it correctly, I think it could be simplified by usage of `mapConcurrently_` and `race`, something along the lines of:

```haskell
firstFoundAddress <- newEmptyTMVarIO
result <- race (atomically $ takeTMVar firstFoundAddress) $ forConcurrently_ addresses $ \addr -> do
  r :: Either E.IOException _ <- try $ cb addr
  for_ r $ atomically . tryPutTMVar firstFoundAddress
either pure $ \() -> throwIO NoAddressFound
```

That's skipping over some of the details, but heads in the direction I was thinking of.

Reply:
It is indeed quite cumbersome, the reasons for which were these

- support the delay between opening sockets (RFC prescribes that and it's just reasonable to not try to connect to all addresses at once)
- minimize the number of threads
- keep the original semantics of "if no success, throw the exception that happened the last", since I don't know what can break if changed (most likely, nothing).

So if you are okay with replacing original connection exceptions with `NoAddressFound` and don't mind creating all `asyncs` at once,  I would test something like 
```
    firstFoundAddress <- newEmptyTMVarIO
    result <- race (atomically $ takeTMVar firstFoundAddress) $ 
                forConcurrently_ (zip addresses [0..]) $ \(addr, n) -> do
        when (n > 0) $ threadDelay $ n * delay
        r :: Either E.IOException _ <- try $ cb addr
        for_ r $ atomically . tryPutTMVar firstFoundAddress
    either pure $ \() -> throwIO NoAddressFound
```

Reply:
I pushed a version that doesn't break any test and is relatively idiomatic. It still creates more threads than possible, but for the most practical cases the difference should be acceptable.

Reply:
I don't mind replacing the exception returned, though using another `TMVar` to capture "first exception raised" should work as well. I'm not sure how many threads we're expecting here. How many addresses would typically be returned?

One very subtle bug that may arise with the code I provided: it's possible that GHC's runtime will mistakenly think there's a deadlock going on. It's probably best to instead have the `TMVar` contain an `Either IOException _` and populate the field after the `forConcurrently_` call.

Reply:
I managed to come up with a version that looks relatively idiomatic, should not deadlock and doesn't break tests on any OS anymore.

> How many addresses would typically be returned?

That is not so easy to say, I would guess that typical host would have 2, but since a lot of people deploy to clouds and use CDNs, they may have quite a few. Here are just some examples from my use case:
```
λ length <$> getAddrInfo (Just hints) (Just "rrdp.ripe.net") (Just "443")
12
λ length <$> getAddrInfo (Just hints) (Just "ripe.net") (Just "443")
2 
λ length <$> getAddrInfo (Just hints) (Just "cc.rg.net") (Just "443")
2
λ length <$> getAddrInfo (Just hints) (Just "rrdp.arin.net") (Just "443")
7
```
The one with a dozen of addresses is the CDN. I would assume that in majority of cases the very first address will be returned before others try to connect, so we'll have a bunch of asyncs sitting in a `threadDelay`. I can come up with something more ugly and efficient with `withAsync` and go-pattern, just to be sure we are not wasting threads.